homeassistant/components/octoprint.py

```diff
@@ -0,0 +1,94 @@
+"""
+Support for monitoring OctoPrint 3D printers.
+
+Uses OctoPrint REST JSON API to query for monitored variables.
+http://docs.octoprint.org/en/master/api/
+"""
+
+import logging
+
+import requests
+
+from homeassistant.components import discovery
+from homeassistant.const import CONF_API_KEY, CONF_HOST
+from homeassistant.helpers import validate_config
+
+DOMAIN = "octoprint"
+OCTOPRINT = None
+
+_LOGGER = logging.getLogger(__name__)
+
+DISCOVER_SENSORS = 'octoprint.sensors'
+DISCOVER_BINARY_SENSORS = 'octoprint.binary_sensor'
+
+
+def setup(hass, config):
+    """Set up OctoPrint API."""
+    if not validate_config(config, {DOMAIN: [CONF_API_KEY],
+                                    DOMAIN: [CONF_HOST]},
+                           _LOGGER):
+        return False
+
+    base_url = config[DOMAIN][CONF_HOST] + "/api/"
+    api_key = config[DOMAIN][CONF_API_KEY]
+
+    global OCTOPRINT
+    try:
+        OCTOPRINT = OctoPrintAPI(base_url, api_key)
+        OCTOPRINT.get("printer")
+    except (requests.exceptions.ConnectionError,
+            requests.exceptions.HTTPError) as conn_err:
+        _LOGGER.error("Error setting up OctoPrint API: %r", conn_err)
+        return False
+
+    for component, discovery_service in (
+            ('sensor', DISCOVER_SENSORS),
+            ('binary_sensor', DISCOVER_BINARY_SENSORS)):
+        discovery.discover(hass, discovery_service, component=component,
+                           hass_config=config)
+
+    return True
+
+
+class OctoPrintAPI(object):
+    """Simple json wrapper for OctoPrint's api."""
+
+    def __init__(self, api_url, key):
+        """Initialize OctoPrint API and set headers needed later."""
+        self.api_url = api_url
+        self.headers = {'content-type': 'application/json',
+                        'X-Api-Key': key}
+        self.printer_last_reading = [{}, None]
+        self.job_last_reading = [{}, None]
+
+    def get(self, endpoint):
```

Comment:
You should add some caching to the different endpoint to make sure that you are not hammering the printer. Right now you will fetch the printer endpoint 4 times if you have sensors and binary sensors up.